iocage_lib/ioc_exec.py

```diff
@@ -83,136 +83,173 @@ def exec_jail(self):
-                                 close_fds=True, bufsize=0, env=self.su_env)
-
-                    # Courtesy of @william-gr
-                    # service(8) and some rc.d scripts have the bad habit of
-                    # exec'ing and never closing stdout/stderr. This makes
-                    # sure we read only enough until the command exits and do
-                    # not wait on the pipe to close on the other end.
-                    #
-                    # Same issue can be demonstrated with:
-                    # $ jexec 1 service postgresql onerestart | cat
-                    # ... <hangs>
-                    # postgresql rc.d command never closes the pipe
-                    rtrn_stdout = b''
-                    rtrn_stderr = b''
-                    for i in ('stdout', 'stderr'):
-                        fileno = getattr(p, i).fileno()
-                        fl = fcntl.fcntl(fileno, fcntl.F_GETFL)
-                        fcntl.fcntl(fileno, fcntl.F_SETFL, fl | os.O_NONBLOCK)
-
-                    timeout = 0.1
-                    while True:
-                        r = select.select([p.stdout.fileno(),
-                                           p.stderr.fileno()], [], [], timeout)[0]
-                        if r:
-                            if p.stdout.fileno() in r:
-                                rtrn_stdout += p.stdout.read()
-                            if p.stderr.fileno() in r:
-                                rtrn_stderr += p.stderr.read()
-
-                        if p.poll() is not None:
-                            if timeout == 0:
-                                break
-                            else:
-                                timeout = 0
-
-                    p.stdout.close()
-                    p.stderr.close()
-
-                    error = True if p.returncode != 0 else False
-
-                    if self.msg_err_return:
-                        return rtrn_stdout, rtrn_stderr, error
-
-                    return rtrn_stdout, error
-                else:
-                    stdout = None if not self.silent else su.DEVNULL
-                    stderr = None if not self.silent else su.DEVNULL
+                cmd = self.command
+
+            if self.msg_return or self.msg_err_return:
+                p = su.Popen(cmd, stdout=su.PIPE, stderr=su.PIPE,
+                             close_fds=True, bufsize=0, env=self.su_env)
+
+                # Courtesy of @william-gr
+                # service(8) and some rc.d scripts have the bad habit of
+                # exec'ing and never closing stdout/stderr. This makes
+                # sure we read only enough until the command exits and do
+                # not wait on the pipe to close on the other end.
+                #
+                # Same issue can be demonstrated with:
+                # $ jexec 1 service postgresql onerestart | cat
+                # ... <hangs>
+                # postgresql rc.d command never closes the pipe
```

Comment:
A good one! I had the same experience with graylog and even had to work with pseudo TTYs to make it start correctly.

Reply:
Seems we have worked around the issue on all ends. But yes, I find it quite unfortunate to rely on a TTY when starting services. We should bring this up and can decide to write a script that tests for similar issues with other services.